tests/cli/shm_ui.c

```diff
@@ -166,38 +173,53 @@ void display_gamepad_state(char** joystick_names, char** button_names) {
     // Read gamepad state if gamepad is connected
     uint32_t pressed_buttons = 0;
     float joystick_vals[4] = {0};
-    if (robot_desc_read(GAMEPAD) == CONNECTED) {
+    int gamepad_connected = (robot_desc_read(GAMEPAD) == CONNECTED);
+    if (gamepad_connected) {
         gamepad_read(&pressed_buttons, joystick_vals);
     } else {
-        mvwprintw(GAMEPAD_WIN, line, 5, "No gamepad connected!");
+        mvwprintw(GAMEPAD_WIN, line, INDENT, "No gamepad connected!");
     }
     wmove(GAMEPAD_WIN, ++line, 0);
 
     // Print joysticks
-    mvwprintw(GAMEPAD_WIN, line++, 1, "Joystick Positions:");
+    wclrtoeol(GAMEPAD_WIN);
+    if (gamepad_connected) {
+        mvwprintw(GAMEPAD_WIN, line, 1, "Joystick Positions:");
+    }
+    wmove(GAMEPAD_WIN, ++line, 0);
     for (int i = 0; i < 4; i++) {
         // Clear previous entry
         wclrtoeol(GAMEPAD_WIN);
         // Display joystick values (This will be 0 if gamepad is not connected)
-        mvwprintw(GAMEPAD_WIN, line, 5, "%s = %f", joystick_names[i], joystick_vals[i]);
+        if (gamepad_connected) {
+            mvwprintw(GAMEPAD_WIN, line, INDENT, "%s\t= %.4f", joystick_names[i], joystick_vals[i]);
+        }
         // Advance line pointer for the next joystick
         wmove(GAMEPAD_WIN, ++line, 0);
     }
 
     // Move to buttons section
-    line += 2;
+    line++;
     wmove(GAMEPAD_WIN, line, 0);
 
     // Print pressed buttons
-    mvwprintw(GAMEPAD_WIN, line++, 1, "Pressed Buttons:");
-    wmove(GAMEPAD_WIN, line, 0);
+    wclrtoeol(GAMEPAD_WIN);
+    if (gamepad_connected) {
+        mvwprintw(GAMEPAD_WIN, line, 1, "Pressed Buttons:");
```

Comment:
is it necessary to go through all the joystick and gamepad logic when the gamepad isn't connected? why can't you just return after the first gamepad check above?

Reply:
I wish it was that easy :P The reason is because at each frame, ncurses draws over the last "snapshot". So lets say all the gamepad stuff was displayed, and now we don't want to display them anymore. We have to explicitly say to clear all that data with whitespace. There are two options here:
1) Clear the entire gamepad window, and draw in the header, box, etc back again
2) Do what I'm doing rn: print iff gamepad is connected

I opted for option 1 for `DEVICE_WIN`. Option 1 has a lot of book-keeping (see that big logic block at the beginning of `display_device_win()`. It's nice to clear a ton of characters at once, but the logic is kinda messy imo and requires global variables.

Option 2 is a bit easier to follow and doesn't require having a global variable. I think this is nicer for smaller windows like `GAMEPAD_WIN`

Reply:
ok i see makes sense that you have to overwrite with whitespace